packages/react/src/index.ts

```diff
@@ -108,17 +115,18 @@ Object.defineProperty(internals.ReactCurrentDispatcher, "current", {
 	set(api) {
 		currentDispatcher = api;
 		if (lock) return;
-		if (api && !isInvalidHookAccessor(api)) {
+		if (lastOwner && api && !isInvalidHookAccessor(api)) {
 			// prevent re-injecting useReducer when the Dispatcher
 			// context changes to run the reducer callback:
 			lock = true;
 			const rerender = api.useReducer(UPDATE, {})[1];
 			lock = false;
-			const currentOwner = internals.ReactCurrentOwner.current;
-			let updater = updaterForComponent.get(currentOwner);
-			if (!updater) {
+
+			let updater = updaterForComponent.get(lastOwner);
+			if (!updater || lastDispatcher !== api) {
 				updater = createUpdater(rerender);
-				updaterForComponent.set(currentOwner, updater);
+				updaterForComponent.set(lastOwner, updater);
+				lastDispatcher = api;
```

Comment:
I haven't looked too much into the core signals mechanism but if signals were only cached after the first run and `signal._updater` was mutated in subsequent runs, that seems to play nice with #70 or #135.

```js
let updater = updaterForComponent.get(lastOwner);
if (!updater) {
  updater = createUpdater(rerender);
  updaterForComponent.set(lastOwner, updater);
} else {
  updater._updater = rerender;
}
```